Add a resource reference generator

ptgott opened the pull request:
Implements RFD 130.

See the included README for information about how the generator works.

This implementation breaks from RFD 130 in the following ways:

- The reference includes separate pages for dynamic resources in order to make it easier to read.
- The generator specifies the kind and version of each dynamic resource. To do this, the generator looks up a method called for each resource type to assign the type's version and kind.
- The generator uses a YAML configuration file. This is because the config ended up including more fields than specified in the RFD, so separating the config from the source made sense to give users less noise to deal with.
- In the example YAML blocks and type table values, leave types empty if we can't process them.

Edit the RFD:

- Be less specific about the output template so this isn't wedded to implementation details.
- Remove the description of example YAML delimiters. These are no longer viable, as we maintain several documentation generators that extract comments from the source.

ptgott committed:
Add a resource reference generator

Closes #16948
Closes #9949

Implements RFD 130.

See the included README for information about how the generator works.

This implementation breaks from RFD 130 in the following ways:

- The reference includes separate pages for dynamic resources in order
  to make it easier to read.
- The generator specifies the kind and version of each dynamic resource.
  To do this, the generator looks up a method called for each resource
  type to assign the type's version and kind.
- The generator uses a YAML configuration file. This is because the
  config ended up including more fields than specified in the RFD, so
  separating the config from the source made sense to give users less
  noise to deal with.
- In the example YAML blocks and type table values, leave types empty if
  we can't process them.

Edit the RFD:

- Be less specific about the output template so this isn't wedded to
  implementation details.
- Remove the description of example YAML delimiters. These are no longer
  viable, as we maintain several documentation generators that extract
  comments from the source.

ptgott commented:
Reviewer note: this PR doesn't include generated docs, since that would make the diff even larger than it is now.

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/main.go:
```diff
@@ -0,0 +1,91 @@
+`
+
+func main() {
+	conf := flag.String("config", "./conf.yaml", configHelp)
```
```suggestion
	conf := flag.String("config", "conf.yaml", configHelp)
```

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,332 @@
+		pc.Resource.Version = verName
+
+		filename := strings.ReplaceAll(strings.ToLower(pc.Resource.SectionName), " ", "-")
+		doc, err := destFS.Create(path.Join(conf.DestinationDirectory, filename+".mdx"))
```
Should this be `filepath.Join`, or does the afero FS take care of that for you?

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,332 @@
+			errs.messages = append(errs.messages, err)
+		}
+
+		if err := template.Must(template.New("Main reference").Parse(referenceTemplate)).Execute(doc, pc); err != nil {
```
I would parse the template once outside of the loop (potentially even in an init func) and then reuse it here. Otherwise we're wasting time re-parsing the same template multiple times.

zmb3 commented on build.assets/tooling/go.mod:
```diff
@@ -69,6 +69,8 @@ require (
 	sigs.k8s.io/yaml v1.4.0 // indirect
 )
 
+require github.com/spf13/afero v1.9.2
```
Move this up into the main require block. (The `go` tool doesn't do this for you sometimes, but you can just move the line manually)

zmb3 reviewed (commented):
Seems reasonable, I'll have to test it out soon!

strideynet reviewed (commented):
First pass looks generally good - glad to see this being worked on!

I think my major concern is there may be parts of this generator that expect/rely on the resource being an old-style resource prior to RFD153 that is generated by GoGo Protobuf. I'd recommend possibly including a new RFD153-compliant resource in your testdata to ensure it'll work with the newest stuff (e.g the Bot resource is one of these)

EdwardDowling commented on build.assets/tooling/cmd/resource-ref-generator/reference/testdata/src/app.go:
```diff
@@ -0,0 +1,88 @@
+	"github.com/gravitational/trace"
+)
+
+var KindApp string = "application"
```
KindApp doesn't seem to be modified, is there a reason it is not a constant?

ptgott replied:
Not sure why this was a `var`. I've changed it to a `const`.

ptgott committed:
Support RFD 153 in the reference generator

RFD 153 resource types include unexported fields and include a
`Metadata` field from a different package than legacy resources types.
They also include the `Metadata` field with a pointer.

Accommodating the different `Metadata` field only requires adding
another entry to `required_field_types` in the config. This change also
adjusts the generator to ignore unexported fields in struct
declarations, and to treat pointer field types the same way as value
field types.

Finally, this change removes some unnecessary spacing from the reference
template.

ptgott committed:
Respond to zmb3 and EdwardDowling feedback

- Embed the reference page template
- Fix comments
- Fix config path in main
- Use filepath.Join instead of path.Join
- Organize go.mod properly
- Move template parsing into an init function
- Change a var declaration to a const

ptgott commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,332 @@
+
+// Intended to be executed with a ReferenceContent.
+// Ampersands are replaced with backticks.
+var referenceTemplate string = strings.ReplaceAll(`---
```
That does make things nicer, thanks!

strideynet commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,332 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
```
```suggestion
// Copyright (C) 2025  Gravitational, Inc.
```

😉

strideynet commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,332 @@
+		return fmt.Errorf("could not parse the configuration file as YAML: %v\n", err)
+	}
+
+	switch {
```
Nit: It's a little unusual to perform validation like this inside of `UnmarshalYAML` but since this isn't production code, I don't think I'm too worried.

strideynet commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,332 @@
+func Generate(srcFS, destFS afero.Fs, conf GeneratorConfig) error {
+	sourceData, err := resource.NewSourceData(srcFS, conf.SourcePath)
+	if err != nil {
+		return fmt.Errorf("can't load Go source files: %v", err)
```
If we use %w, then the error is properly wrapped.

```suggestion
		return fmt.Errorf("loading Go source files: %w", err)
```

ptgott committed:
Incorporate strideynet feedback

- Update copyright year.
- Wrap errors properly in `fmt.Errorf`.

ptgott committed:
Edit descriptions in generated resource references

Enclose curly braces in backticks to prevent docs builds from breaking.

ptgott commented:
Fixed another bug in https://github.com/gravitational/teleport/pull/50515/commits/b5d2ec42904fb9ec92fbfe8d03f16d9459788f3b.

Opened a stacked PR to show the finished result in the Amplify preview: https://github.com/gravitational/teleport/pull/51506

ptgott committed:
Address linter issues

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/README.md:
````diff
@@ -0,0 +1,81 @@
+## Usage
+
+```
+$ cd build.assets/tooling/cmd/resource-ref-generator
+$ go run . -config config.yaml
````
Consider adding a make target to the root makefile.

ptgott replied:
So far, the root Makefile has targets for checking that auto-generated docs are up to date ([here](https://github.com/gravitational/teleport/blob/d15466c5371fd8a2456932891afbffe0b2a67cb6/Makefile#L1606-L1628)), but not for generating the docs. I could follow this convention and put the target in [`build.assets/Makefile`](https://github.com/gravitational/teleport/blob/master/build.assets/Makefile), though a comment in the file says this is for producing official Teleport releases. At the same time, a root make target for running the docs generators would be nice, since they're spread throughout the source.

Come to think of it, is `build.assets/tooling/cmd` the best place for this project? Would it make sense to place it in `docs`?

ptgott replied:
Added

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/main.go:
```diff
@@ -0,0 +1,92 @@
+		os.Exit(1)
+	}
+
+	osFS := afero.NewOsFs()
```
Does the standard library `fs.FS` work here instead of using afero?

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.tmpl:
```diff
@@ -0,0 +1,53 @@
+
+{{ .Resource.Description }}
+
+{/*Automatically generated from: {{ .Resource.SourcePath}}*/}
```
Should we move this line to the top of the file and make it a bit louder so people don't unintentionally try to edit this file?

I'm not sure if gopls operates on .tmpl files, but if it does there's a standard "code generated" comment that will actually tell editors to warn users who try to edit generated files.

zmb3 replied:
This is what I was thinking of: https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

> To convey to humans and machine tools that code is generated, generated source should have a line that matches the following regular expression (in Go syntax):
>
> ```
> ^// Code generated .* DO NOT EDIT\.$
> ```

ptgott replied:
Thanks! The generator adds the "DO NOT EDIT" line to the top of `.mdx` files, so it shouldn't need to match the pattern intended for autogenerated Go files.

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference_test.go:
```diff
@@ -0,0 +1,358 @@
+// files, delete the destination directory (reference/testdata/dest) and run the
+// test again.
+func TestGenerate(t *testing.T) {
+	// Define paths based on relative paths from this Go source file to
+	// guarantee consistent results regardless of where a user runs "go test".
+	_, callerPath, _, _ := runtime.Caller(0)
```
You shouldn't need to do this. `go test` guarantees that the current working directory will be the directory containing the test (`reference` in this case).

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference_test.go:
```diff
@@ -0,0 +1,358 @@
+	// Define paths based on relative paths from this Go source file to
+	// guarantee consistent results regardless of where a user runs "go test".
+	_, callerPath, _, _ := runtime.Caller(0)
+	fmt.Println(callerPath)
```
```suggestion
```

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/testdata/src/app.go:
```diff
@@ -0,0 +1,90 @@
+package typestest
+
+import (
+	fmt "fmt"
```
Why?

ptgott replied:
We use `fmt` in `api/types/app.go`, which this file is based on. Not sure where that strange package alias came from.

zmb3 commented on build.assets/tooling/go.sum:
This is quite a lot of new dependencies. I guess most of them must come from afero?

ptgott committed:
Respond to zmb3 feedback

- Use standard library packages instead of afero for storing generated
  files in tests. As part of this work, change the `SourcePath` in
  `ReferenceEntry` to use a directory relative to the project root,
  meaning that we don't need to include a temporary directory path in
  test expectations. Also clarify some errors in `Generate` that were
  returned while debugging in-test directory management.
- Remove unnecessary code to find the caller of `TestGenerate`.
- Remove unnecessary package alias.
- Move auto-generation warning comments to the top of each generated
  page. (These were formerly at the top of each section, since an
  earlier iteration attempted to generate a single reference for all
  resources.)

ptgott committed:
Add a Make target

...for generating dynamic resource reference documentation

ptgott commented:
@zmb3 Putting this back on your radar for when you have some time to look at it again. Thanks!

ptgott committed:
Fix linter errors

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+
+	// If the configured excluded resources include this type declaration,
+	// don't process it.
+	for _, r := range excludedResources {
```
Consider `slices.ContainsFunc`.

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+
+	// Compare the types of fields in the struct type with the required
+	// fields types. Only one required field type must be present.
+	var m bool
```
What does `m` stand for?

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+	// fields types. Only one required field type must be present.
+	var m bool
+	for _, fld := range str.Fields.List {
+		if len(fld.Names) != 1 {
+			continue
+		}
```
This sill skip over things like:

```
type Foo struct {
    Bar, Baz int
}
```

Again, it's probably a safe assumption to make, but I wonder if we should at least log/warn about things that don't match our expectations rather than treating them like they don't exist.

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+		// comes from the same package as the outer struct type, then
+		// assign a package name depending on the expression used to
+		// declare the field type.
+		gopkg := d.PackageName
+		pi := getPackageInfoFromExpr(fld.Type)
+		if pi.PackageName != "" {
+			gopkg = pi.PackageName
+		}
```
```suggestion
		pi := getPackageInfoFromExpr(fld.Type)
		gopkg := cmp.Or(pi.PackageName, d.PackageName)
```

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+			gopkg = pi.PackageName
+		}
+
+		for _, ti := range finalTypes {
```
`slices.ContainsFunc` here?

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+
+	// Extract data from a declaration to transform it into a reference
+	// entry later
+	errs := GenerationError{messages: []error{}}
```
```suggestion
	var errs GenerationError
```

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,306 @@
+
+		filename := strings.ReplaceAll(strings.ToLower(pc.Resource.SectionName), " ", "-")
+		docpath := filepath.Join(conf.DestinationDirectory, filename+".mdx")
+		doc, err := os.Create(docpath)
```
Don't we need to close the file when we're done writing to it?

russjones commented:
@ptgott What is the state of this RFD? Do we plan to implement?

ptgott commented:
> @ptgott What is the state of this RFD? Do we plan to implement?

I intend to get back to this project, but it is not a goal for this quarter. I am waiting for time to open up.
